### ISSUE TITLE
Fully tested production ready bridge

### ISSUE THREAD

**Issue by nearmax** [near/rainbow-bridge#50]
Here I suggest we discuss the design of the rainbow birdge devops.

# Command line interface

We need to have a CLI to execute the following commands.
## User commands
- [x] `transfer-fun-eth2near` -- transfers from fungible token on ETH to an equivalent fungible token on NEAR blockchain;
- [x] `transfer-fun-near2eth` -- transfers from fungible token on NEAR to an equivalent fungible token on ETH blockchain;
- [ ] `transfer-eth-eth2near` -- transfers ETH from ETH Blockchain to a fungible token on NEAR blockchain;
- [ ] `transfer-near-eth2near` -- transfers from fungible token representing NEAR on ETH Blockchain to NEAR on NEAR blockchain;
- [ ] `transfer-eth-near2eth` -- transfers from fungible token representing ETH on NEAR blockchain to ETH on ETH blockchain;
- [ ] `transfer-near-near2eth` -- transfers NEAR from NEAR blockchain to a fungible token on ETH blockchain;
- [ ]  `dump-eth-proof` -- dumps the proof of an ETH event to local JSON file;
- [ ]  `verify-eth-proof` -- verifies the proof of an ETH event using Eth2NearProver contract on NEAR blockchain;

These commands can be used for various kinds of transfers.

## Maintainer commands
- [x] `init-near-contracts` -- deploys and initializes Eth2NearClient, Eth2NearProver contracts to NEAR blockchain;
- [x] `init-eth-contracts` -- deploys and initalizes Near2EthClient, Near2EthProver, EthLocker contracts to ETH blockchain;
- [ ] `verify-near-header` -- will download the NEAR header that was recently submitted to Near2EthClient contract on ETH blockchain and verify it. Will output true or false depending on whether the verification succeeds or not;

## Testing commands
- [x] `dump-eth-headers` -- dumps a provided range of ETH headers to a local file so that they can be used to perform CI tests on the Eth2NearClient contract. The range can specify how many  headers starting from the most recent one should be dumped;
- [x] `dump-near-headers` -- dumps a provided range of NEAR light headers to a local file so that can be used to perform CI tests on the Near2EthClient contract (while the contract does not verify signatures of the headers, it still verifies everything else in the header, like hashes);
- [x] `dump-eth-proofs` -- dumps proofs to all events in the provided range of ETH headers. The dump can be used to perform tests on the Eth2NearProver contract;
- [x] `dump-near-proofs` --dumps proofs to all transaction outcomes in the provided range of NEAR headers. The dump can be used to perform tests on Near2EthProver contract.
- [x] `init-near-test-contracts` -- deploys and initializes fungible token on NEAR blockchain;
- [x] `init-eth-test-contracts` -- deploys and initializes ERC20 and Emitter to ETH blockchain;

# Services
The following services can be started either using pm2 or docker compose.
## Production services
- [x] `eth2near-relay` -- a service that listens to new ETH headers and submits them to NEAR blockchain;
- [x] `near2eth-relay` -- a service that extracts `LightClientBlockView` every X minutes (X depends on the bond timing specified in Near2EthClient) and submits it to Near2EthClient;
- [x] `near2eth-client-watchdog` -- a service that monitors headers that were submitted to Near2EthClient, verifies the signatures in them, and if any signature is incorrect submits a challenge.

## Development services
Near node and Ganache.

## Testing services
- [ ] `eth2near-fuzz-tester` -- a service that would send a bunch of calls to Emitter contract deployed to local Ganache and verify all of them on local Near node;
- [ ] `near2eth-fuzz-tester` -- a service that would send a bunch of calls to Emitter contract deployed to local Near node and verify all of them on local Ganache.
- [ ] `bridge-back-n-forth`  -- a service that would move a bunch of fungible tokens back and forth between Ethereum and Near nodes.

@k06a , @ailisp , @chefsale WDYT?

**Comment by ailisp** [comment]
@nearmax What is still missing currently for this Epic?

**Comment by nearmax** [comment]
@ailisp Sorry for the delay, I have update the description with the checkboxes.

**Comment by ailisp** [comment]
@nearmax Looks cool!
>  dump-eth-headers -- dumps a provided range of ETH headers to a local file so that they can be used to perform CI tests on the Eth2NearClient contract. The range can specify how many headers starting from the most recent one should be dumped;

So i'll use this command to trigger "download blocks from ETH" #58 ?

**Comment by ailisp** [comment]
>  transfer-fun-eth2near -- transfers from fungible token on ETH to an equivalent fungible token on NEAR blockchain;
transfer-fun-near2eth -- transfers from fungible token on NEAR to an equivalent fungible token on ETH blockchain;
transfer-eth-eth2near -- transfers ETH from ETH Blockchain to a fungible token on NEAR blockchain;
transfer-near-eth2near -- transfers from fungible token representing NEAR on ETH Blockchain to NEAR on NEAR blockchain;
transfer-eth-near2eth -- transfers from fungible token representing ETH on NEAR blockchain to ETH on ETH blockchain;
transfer-near-near2eth -- transfers NEAR from NEAR blockchain to a fungible token on ETH blockchain;

These I think more friendly to be a single transfer command with different parameters

**Comment by nearmax** [comment]
@ailisp 

> So i'll use this command to trigger "download blocks from ETH" #58 ?

Yes, please.

> These I think more friendly to be a single transfer command with different parameters

They will have different arguments. To my understanding that's what commands are used for -- to separate different sets of arguments, but I don't hold strong opinion on it.

**Comment by chefsale** [comment]
I finished prepare and clean methods, working now on the start which would run PM2 for managing the processes, as well plan to implement the stop command proposed by @ailisp . This makes a lot of sense to be added I think (https://github.com/near/rainbow-bridge/issues/43). Once that is done, @nearmax let me know what else do you need help with.

**Comment by ailisp** [comment]
@nearmax Do you want to make a quick update to current commands?

**Comment by nearmax** [comment]
@ailisp I think we have implemented most of the important stuff, the remaining things need to be revisited. Specifically, many of the remaining user commands listed above are outside the current scope and would need a separate issue or even an Epic for each of them, because they might have a lot of work items.

Testing services can be extracted into a separate issue. Therefore I am closing this issue.
